### PR TITLE
Use custom parser for gts/gjs

### LINKED ISSUE #1659: Conflicts with prettier-plugin-ember-template-tag + eslint-plugin-prettier when linting gjs/gts files
I had a bug filed on my new project regarding an error being thrown when run via eslint + eslint-plugin-ember + eslint-plugin-prettier: https://github.com/gitKrystan/prettier-plugin-ember-template-tag/issues/20

The further I delve into the issue, the more I think I might need help from this team to solve it, and that possibly the solution might involve changes to this repo rather than mine.

Repro: https://github.com/gitKrystan/simple-gjs-project/pull/2

This is the error thrown:

```
npm run lint:js                             

> gjs-project@0.0.0 lint:js
> eslint . --cache


Oops! Something went wrong! :(

ESLint: 7.32.0

SyntaxError: Invalid regular expression: /\b[__GLIMMER_\b/: Unterminated character class
    at new RegExp (<anonymous>)
    at /Users/krystanhuffmenne/Code/gjs-project/node_modules/eslint-plugin-ember/lib/preprocessors/glimmer.js:121:34
    at Array.map (<anonymous>)
    at mapRange (/Users/krystanhuffmenne/Code/gjs-project/node_modules/eslint-plugin-ember/lib/preprocessors/glimmer.js:106:20)
    at Linter._verifyWithProcessor (/Users/krystanhuffmenne/Code/gjs-project/node_modules/eslint/lib/linter/linter.js:1339:16)
    at Linter._verifyWithConfigArray (/Users/krystanhuffmenne/Code/gjs-project/node_modules/eslint/lib/linter/linter.js:1273:25)
    at Linter.verify (/Users/krystanhuffmenne/Code/gjs-project/node_modules/eslint/lib/linter/linter.js:1235:25)
    at Linter.verifyAndFix (/Users/krystanhuffmenne/Code/gjs-project/node_modules/eslint/lib/linter/linter.js:1428:29)
    at verifyText (/Users/krystanhuffmenne/Code/gjs-project/node_modules/eslint/lib/cli-engine/cli-engine.js:240:48)
    at CLIEngine.executeOnFiles (/Users/krystanhuffmenne/Code/gjs-project/node_modules/eslint/lib/cli-engine/cli-engine.js:808:28)
```

First of all, it seems like there should be some escaping happening here:
https://github.com/ember-cli/eslint-plugin-ember/blob/db745a3881686d50e486be67d6851d806ac770d6/lib/preprocessors/glimmer.js#L120

Because eslint-plugin-prettier runs Prettier as an ESLint rule, my plugin causes the Prettier ESLint rule to log the following violation message:

```
{
  ruleId: "prettier/prettier",
  severity: 2,
  message: "Replace `[__GLIMMER_TEMPLATE(`⏎··<h1>···Hello·{{@who}}.·</h1>⏎`,·{·strictMode:·true·})];` with `<template><h1>·Hello·{{@who}}.·</h1></template>`",
  line: 1,
  column: 16,
  nodeType: null,
  messageId: "replace",
  endLine: 3,
  endColumn: 27,
}
```

Interestingly, the message has the preprocessed `[__GLIMMER_TEMPLATE(` shenanigans when it probably should not. (Not sure yet if that because of something I am doing or because of something happening in this package).

As a result, the line linked above is generating `new RegExp('\b[__GLIMMER_\b')` which indeed has an opening `[` that the RegExp constructor interprets as the beginning of a character set.

I've noticed that this doesn't _always_ happen. For example, if I use the `<template>` tag in the "top-level class" position, `npm run lint:js` works properly (ish..it still has `[__GLIMMER_TEMPLATE()]` in the message, and I think the line/column numbers might be off):

```
// works-ish.gjs

import Component from '@glimmer/component';

export default class MyComponent
  extends Component {

        <template>


    <h1>   Hello {{@who}}. </h1>
  </template>
}
```

```
npm run lint:js 

> gjs-project@0.0.0 lint:js
> eslint . --cache

Debugger attached.

/Users/krystanhuffmenne/Code/gjs-project/app/components/broken.gjs
  2:1  error  Delete `⏎`                                                                                                                                                        prettier/prettier
  2:1  error  Delete `⏎·`                                                                                                                                                       prettier/prettier
  2:1  error  Replace `⏎········[__GLIMMER_TEMPLATE(`⏎⏎⏎····<h1>···Hello·{{@who}}.·</h1>⏎··`,·{·strictMode:·true·})]` with `··<template><h1>·Hello·{{@who}}.·</h1></template>`  prettier/prettier

✖ 3 problems (3 errors, 0 warnings)
```

### LINKED ISSUE #1683: Glimmer Processor can't handle errors without lines
https://github.com/ember-cli/eslint-plugin-ember/blob/11556154cb026ffc58c521a778a8f776fa62f354/lib/preprocessors/glimmer.js#L107

Some errors do not have lines associated with them, so this does not work.

### LINKED ISSUE #1747: False positive `no-undef` for template-imports and contextual components
This is failing linting with `form` and `field` "locals" (they are not local to the outer template scope, so _not_ supposed to be provided by JS scope, which ESLint thinks here) falsely violating `no-undef`:

```gjs
<template>
  <HeadlessForm @data={{data}} as |form|>
    <form.field @name="firstName" as |field|>
      <field.label>First Name</field.label>
      <field.input data-test-first-name />
    </form.field>
    <form.field @name="lastName" as |field|>
      <field.label>Last Name</field.label>
      <field.input data-test-last-name />
    </form.field>
  </HeadlessForm>
</template>
```

I think I have identified the root cause in ` @glimmer/syntax`, keeping this for tracking. /cc @NullVoxPopuli

### LINKED ISSUE #1770: Cannot do type-aware lints with gts files
Initially, I was prompted to set `parserOptions.extraFileExtensions`. I did this and then received the following error:

```
{
  message: {
    ruleId: null,
    fatal: true,
    severity: 2,
    message: 'Parsing error: ESLint was configured to run on `<tsconfigRootDir>/addons/luna/addon/components/util/cache-state.gts/0_/Users/pwagenet/Development/AuditBoard/auditboard-frontend/addons/luna/addon/components/util/cache-state.gts/0_/Users/pwagenet/Development/AuditBoard/auditboard-frontend/addons/luna/addon/components/util/cache-state.gts/0_/Users/pwagenet/Development/AuditBoard/auditboard-frontend/addons/luna/addon/components/util/cache-state.gts` using `parserOptions.project`: <tsconfigRootDir>/../../../../../users/pwagenet/development/auditboard/auditboard-frontend/tsconfig.json\n' +
      'However, that TSConfig does not include this file. Either:\n' +
      "- Change ESLint's list of included files to not include this file\n" +
      '- Change that TSConfig to include this file\n' +
      '- Create a new TSConfig that includes this file and include it in your parserOptions.project\n' +
      'See the typescript-eslint docs for more info: https://typescript-eslint.io/linting/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file',
    line: undefined,
    column: undefined
  }
}
```

Unfortunately, it's not possible to add unrecognized file extensions to the tsconfig.

(Note I had to edit the source to log the error due to the missing `line`. See also #1667.)

### LINKED ISSUE #1894: <template> tag parsing issue in templates shorter than 3 characters
1. `ember new --yarn zomg`, as of now that gives you:
   1. `ember-cli@5.0.0`
   1. `eslint@8.43.0`
   1. `eslint-plugin-ember@11.8.0`
1. Disable prettier integration (to isolate the problem to eslint):
   1. Remove `eslint-config-prettier` package
   1. Remove `eslint-plugin-prettier` package
   1. Remove `extends` `plugin:prettier/recommended`
   1. `yarn`
1. Add `app/components/foo.gjs` with the following content
1. Run `yarn lint:js` (which is `eslint . --cache`)

# 0 characters

```
export default <template></template>;
```

```
/private/tmp/zomg/app/components/foo.gjs
  0:0  error  Preprocessing error: Cannot split a chunk that has already been edited (1:13 – "</template>")

✖ 1 problem (1 error, 0 warnings)
```

# 1 character

```
export default <template>h</template>;
```

```
/private/tmp/zomg/app/components/foo.gjs
  0:0  error  Preprocessing error: Cannot overwrite across a split point

✖ 1 problem (1 error, 0 warnings)
```

# 2 characters

```
export default <template>hi</template>;
```

```
/private/tmp/zomg/app/components/foo.gjs
  0:0  error  Preprocessing error: Cannot overwrite a zero-length range – use appendLeft or prependRight instead

✖ 1 problem (1 error, 0 warnings)
```

# 3 characters

```
export default <template>hi!</template>;
```

```
✨  Done in 0.59s.
```

(The result is the same if you put the template tag inside a class body)

### PR DESCRIPTION
**Pull request by patricklx**
I was writing this parser for eslint-plugin-template-lint,
but I think it makes sense to have it for the ember-plugin as well.
Also, as preprocess changes the filename to a virtual one, type aware linting will not work: https://github.com/typescript-eslint/typescript-eslint/issues/6177#issuecomment-1341609701
By using a parser, that would be fixed
bonus:
 * prettier eslint plugin (with template tag prettier plugin) will just work for gts/gjs
 * semi rule works same as prettier tag plugin
 * enables type aware lints
 * can detect unused block params in templates
 * can detect undef vars in PathExpression
 * can add eslint directive comments in mustache or html 
 
 disadvantage:
* prettier will not work without template tag prettier plugin for gts/gjs files

fixes #1770 
fixes #1894 
fixes #1747 
fixes #1683 
fixes #1659 
also fixes https://github.com/gitKrystan/prettier-plugin-ember-template-tag/issues/81
fixes https://github.com/gitKrystan/prettier-plugin-ember-template-tag/issues/84
closes #1667

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -17,7 +17,7 @@ jobs:
     strategy:
       matrix:
         os: [ ubuntu, windows ]
-        node-version: [14.x, 16.x, 18.x]
```
Comment (patricklx) [inline .github/workflows/ci.yml:20]: because of @typescript-eslint/scope-manager

File: package.json [hunk 2]
```diff
@@ -65,25 +65,25 @@
...
+    "@babel/eslint-parser": "^7.15.8",
     "@ember-data/rfc395-data": "^0.0.4",
     "@glimmer/syntax": "^0.84.2",
+    "@typescript-eslint/parser": "^5.31.0",
+    "@typescript-eslint/scope-manager": "^6.2.0",
     "css-tree": "^2.0.4",
     "ember-rfc176-data": "^0.3.15",
-    "ember-template-imports": "^3.4.2",
-    "ember-template-recast": "^6.1.4",
+    "ember-template-tag": "^2.2.3",
```
Comment (NullVoxPopuli) [inline package.json:75]: any reason to not use `content-tag`?

Comment (patricklx) [reply package.json:75]: content-tag only transforms the content. it does not provide any info about the changes done.
It also does not guarantee the same code syntax & spacing, as its not a recast and also adds import statement at the top, which changes all lines.
i tried also to use source mapping, but it doesn't matter if the syntax is changed.

**Comment by patricklx** [comment]
I don't think we can use content-tag unless it transforms a file into a useable form for linting.

**Comment by patricklx** [comment]
Anyone else who can review? @bmish ?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (NullVoxPopuli) [reply .github/workflows/ci.yml:20]: This'll require a major release to drop node 14, cc @bmish

File: lib/parsers/gjs-parser.js [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+      ? typescriptParser.parseForESLint(jsCode, { ...options, ranges: true })
+      : babelParser.parseForESLint(jsCode, { ...options, ranges: true });
+    if (!info.replacements?.length) {
+      return result;
+    }
+    const preprocessedResult = preprocessGlimmerTemplates(info, code);
+    const { templateVisitorKeys } = preprocessedResult;
+    const visitorKeys = { ...result.visitorKeys, ...templateVisitorKeys };
+    result.isTypescript = isTypescript;
+    convertAst(result, preprocessedResult, visitorKeys);
```
Comment (NullVoxPopuli) [inline lib/parsers/gjs-parser.js:427]: does this mean that the template-lint rules could be added as eslint rules? I saw you had `@glimmer/syntax` as well?

File: lib/parsers/gjs-parser.js [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+ * AST.
+ * This also creates the scopes for the Glimmer Blocks and registers the block params
+ * in the scope, and also any usages of variables in path expressions
+ * this allows the basic eslint rules no-undef and no-unsused to work also for the
+ * templates without needing any custom rules
+ * @param result
+ * @param preprocessedResult
+ * @param visitorKeys
+ */
+function convertAst(result, preprocessedResult, visitorKeys) {
```
Comment (NullVoxPopuli) [inline lib/parsers/gjs-parser.js:323]: why is all this code needed? 
do we need to zip in the glimmer AST with the JS AST?
does doing so allow us to eventually phase out template-lint? (or adopt the rules in to here?)

File: lib/rules/no-unused-services.js [hunk 5]
```diff
@@ -380,6 +363,13 @@ module.exports = {
           currentClass.uses.add(name);
         }
       },
+      GlimmerPathExpression(node) {
```
Comment (NullVoxPopuli) [inline lib/rules/no-unused-services.js:366]: this _is_ much nicer!

File: package.json [hunk 6]
```diff
@@ -65,25 +65,25 @@
...
   },
   "dependencies": {
+    "@babel/eslint-parser": "^7.15.8",
     "@ember-data/rfc395-data": "^0.0.4",
     "@glimmer/syntax": "^0.84.2",
+    "@typescript-eslint/parser": "^5.31.0",
+    "@typescript-eslint/scope-manager": "^6.2.0",
     "css-tree": "^2.0.4",
     "ember-rfc176-data": "^0.3.15",
-    "ember-template-imports": "^3.4.2",
```
Comment (NullVoxPopuli) [inline package.json:72]: good to see this gone

File: tests/lib/rules-preprocessor/gjs-gts-processor-test.js [hunk 7]
```diff
@@ -472,6 +499,40 @@ describe('lint errors on the exact line as the <template> tag', () => {
   });
 });
 
+describe('supports eslint directives inside templtes', () => {
```
Comment (NullVoxPopuli) [inline tests/lib/rules-preprocessor/gjs-gts-processor-test.js:502]:
```suggestion
describe('supports eslint directives inside templates', () => {
```

File: lib/index.js [hunk 8]
```diff
@@ -2,18 +2,10 @@
...
 
-const gjs = require('./preprocessors/glimmer');
-
 module.exports = {
   rules: requireIndex(`${__dirname}/rules`),
   configs: requireIndex(`${__dirname}/config`),
   utils: {
     ember: require('./utils/ember'),
   },
-  processors: {
```
Comment (NullVoxPopuli) [inline lib/index.js:13]: we still probably want to export the parser -- @chancancode had a use case where the (now old) parser breaks on old code -- it'd be good for us to provide a way for folks to configure their own overrides in the lint config and set the parser:
```cjs
overrides: [
  {
    files: ['modern/**/*.gts'],
    parser: require('eslint-plugin-ember/gts-parser'),
    // ...
  }
]
```

Comment (NullVoxPopuli) [reply lib/index.js:13]: (we'd need to use `package.json#exports` to expose this tho -- maybe for another PR)

**Review by NullVoxPopuli — APPROVED** [review]

File: lib/parsers/gjs-parser.js (hunk 3 above)
Comment (patricklx) [reply lib/parsers/gjs-parser.js:427]: yes, I'm already having most of them working with this parser over at eslint-plugin-template-lint. running them natively.

File: lib/parsers/gjs-parser.js (hunk 4 above)
Comment (patricklx) [reply lib/parsers/gjs-parser.js:323]: yes, possibly. but mostly now to auto-detect unused and undef vars

File: tests/lib/rules-preprocessor/gjs-gts-processor-test.js (hunk 7 above)
Comment (patricklx) [reply tests/lib/rules-preprocessor/gjs-gts-processor-test.js]: fixed

File: .github/workflows/ci.yml (hunk 1 above)
Comment (patricklx) [reply .github/workflows/ci.yml:20]: I can also try to downgrade to 5.62.0.
They dropped 14 with the 6.0.0 release

**Comment by patricklx** [comment]
Also note that the parser can be globally overwritten just by specifying it in the eslintrc, which can cause problems.
https://github.com/patricklx/eslint-plugin-ember-template-lint#important

File: lib/index.js (hunk 8 above)
Comment (patricklx) [reply lib/index.js:13]: Or rather have an option to specify the internal base parser. 
Now I'm just choosing between Babel parser and typescript eslint parser. But that could also be an option passed to the gts parser

**Comment by patricklx** [comment]
added typescript typed eslint tests

File: .github/workflows/ci.yml (hunk 1 above)
Comment (patricklx) [reply .github/workflows/ci.yml]: i downgraded as i also needed lower version for testing with typed eslint

File: tests/lib/rules-preprocessor/my-component.gts
Comment (NullVoxPopuli) [inline tests/lib/rules-preprocessor/my-component.gts:1]: Empty file?

File: tests/lib/rules-preprocessor/gjs-gts-processor-test.js [hunk 9]
```diff
@@ -27,13 +30,28 @@ function initESLint(parser = '@babel/eslint-parser') {
       env: {
         browser: true,
       },
-      parser,
       parserOptions: {
```
Comment (NullVoxPopuli) [inline tests/lib/rules-preprocessor/gjs-gts-processor-test.js:33]: Oh i see how those files are used legit

**Review by NullVoxPopuli — APPROVED** [review]

**Review by bmish — APPROVED** [review]
Thanks a lot for this!

**Comment by Techn1x** [comment]
Bumped to this and I seem to get an eslint error at the very start of all of my `<template>` tags now 🤔 

Screenshot 2023-08-22 at 8 23 30 pm

This is defined within a class
```ts
export class Alert extends Component<AlertSignature> {
  <template> ... </template>
}
```

**Comment by patricklx** [comment]
@Techn1x do you have specifies a parser in your eslint config?  Like the typescript parser?
If so, then remove it and extend from `'plugin:@typescript-eslint/base'`. Thag will set it up correctly and not interfere with emberjs parser

**Comment by bmish** [comment]
@Techn1x can you open a separate issue to discuss this? @patricklx if this is a breaking change, we will need to revert it.

**Comment by Techn1x** [comment]
Thanks, will do.

**Pull request merged**